[6.2.z]implement shared function

ldjebran opened the pull request:
close issue: https://github.com/SatelliteQE/robottelo/issues/4442 

Shared function is a decorator, that enable a function once called to store
the results to storage, any ulterior call from the same or other processes will
return the stored results, which make the shared function results persistent.

Note: Shared function store it's data as json. The results of the decorated
    function must be json compatible.

Usage::

```python
    from robottelo.decorators.func_shared.shared import shared

    @shared
    def module_level_shared(*args, **kwargs):
        # do some
        # the result of a shared function must be json compatible
        return any_data

    class SomeTestCase1(TestCase):

        @shared
        def _shared_function(cls):

            org = make_org()
            # upload manifest
            repo = make_repository()
            return {'org_id': org['id'], 'repo_id': repo['id']}

        @classmethod
        @shared
        def setUpClass(cls):

            data = cls._shared_function()
            other_data = module_level_shared()

            cls.org = Org.info({'id': data['org_id']})
            cls.repo = Repository.info({'id': data['repo_id']})
            return

    # the shared function can be called an other time to be able to initiate
    # specific data

    class SomeTestCase2(TestCase):

        @classmethod
        @shared(inject=True, injected_kw='_injected')
        def setUpClass(cls, org_id=None, repo_id=None, _injected=False):

            if _injected:
                cls.org = Org.info({'id': org_id})
                cls.repo = Repository.info({'id': repo_id})
            else:
                # create the org
                cls.org = make_org()
                # upload manifest
                cls.repo = make_repository()

            # create a virtual machine

            # shared function with injected=True, must return a dict
            # the resulting dictionary will be injected in other calls as
            # kwargs, an added bool kw argument by default named _injected
            # should be added to the function kwargs, to be able to be notified
            # that the kwargs are injected from already stored result
            return {'org_id': cls.org['id'], 'repo_id': cls.repo['id']}

        # in case we do not want the injected key word in kwargs
        # simply , declare injected_kw=None
        @classmethod
        @shared(inject=True, injected_kw=None)
        def shared_class_method(cls, org_id=None, repo_id=None):
             if org_id is not None:
                cls.org = Org.info({'id': org_id})
             else:
                # create the org
                cls.org = make_org()
                # upload manifest
             if repo_id is not None:
                cls.repo = Repository.info({'id': repo_id})
             else:
                cls.repo = make_repository()

            # create a virtual machine

            return {'org_id': cls.org['id'], 'repo_id': cls.repo['id']}
```
Features : 

* main functionality unittest covered
* supported storage: file or redis 
* restored failures 
* retry call (in case of first call failed) can be configured, by default  retries=2
* stored data timeout (by default 24 hours)
* when using file storage, cmd: "make clean-shared" to clean the temp storage files
* can share function results at  module level , classes up to second level
* the shared results can be shared between processes and with redis usage between processes running 
on different hosts.

Example usage:
1- we have a capsule setup, but to create the capsule org with it's content views that contain the needed  rh products needed by the capsule to install, it takes many time, we can create a shared setup_capsule_org function that will be used by the setup_capsule if no org indicated, that way all the capsule tests will use the same capsule org (as the capsule org content is created when first called). 
2- any setUpClass that need to share data with all the tests processes.
3- We can now use shared setUpClass  with publish manifest without run in one thread decorator (but need's the locking of the manifest upload).

Note: the result of shared function is shared also between all tiers, if they are started before the result timeout .

Inconvenient:
* We can share only the data that are json persistent
  for the moment CLI info results can be shared directly as they are dictionaries, nailgun entities needs some work to be  done, but enties can be passed by ids and restored for example: 

      cls.org = Org.info({'id': org_id}) 

      or 

      entities.Organization(id=org_id).read()

for the moment if any key value return by a shared function is an entity , the the saved value will be 
value.to_json_dict() (look at api testcase on how nailgun entities are restored)

Example test done:
cli.test.host.HostCreateTestCase
**Note test is not included in this PR as think that PR must be merged alone**
tests done with 8 processes
```python
class HostCreateTestCase(CLITestCase):
    """Tests for creating the hosts via CLI."""

    @classmethod
    @shared(inject=True)
    def setUpClass(cls, new_org=None, puppet_cv=None, puppet_env=None,
                   puppet_class=None, _injected=False):
        """Create organization, lifecycle environment, content view, publish
        and promote new version to re-use in tests.
        """
        super(HostCreateTestCase, cls).setUpClass()

        if _injected:
            cls.new_org = new_org
            cls.puppet_cv = puppet_cv
            cls.puppet_env = puppet_env
            cls.puppet_class = puppet_class
        else:
            cls.new_org = make_org()
            # Setup for puppet class related tests
            puppet_modules = [
                {'author': 'robottelo', 'name': 'generic_1'},
            ]
            cls.puppet_cv = publish_puppet_module(
                puppet_modules, CUSTOM_PUPPET_REPO, cls.new_org['id'])
            cls.puppet_env = Environment.list({
                'search': u'content_view="{0}"'.format(
                    cls.puppet_cv['name'])})[0]
            cls.puppet_class = Puppet.info({
                'name': puppet_modules[0]['name'],
                'environment': cls.puppet_env['name'],
            })
        return dict(new_org=cls.new_org, puppet_cv=cls.puppet_cv,
                    puppet_env=cls.puppet_env, puppet_class=cls.puppet_class)
```
The stored shared data file:
http://pastebin.test.redhat.com/468164 

test run with sat-6.2.8-4.1b as the latest version seems to fail on that tests.

test log:
```
(sat-6.2.z) dlezz@elysion:~/projects/robottelo-fork$ py.test tests/foreman/cli/test_host.py -n 8 -v -k "HostCreateTestCase"
================================================= test session starts ==================================================
platform linux2 -- Python 2.7.13, pytest-2.9.2, py-1.4.32, pluggy-0.3.1 -- /home/dlezz/.pyenv/versions/sat-6.2.z/bin/python
cachedir: .cache
rootdir: /home/dlezz/projects/robottelo-fork, inifile: 
plugins: xdist-1.15.0, services-1.1.14, cov-2.3.1
[gw0] linux2 Python 2.7.13 cwd: /home/dlezz/projects/robottelo-fork
[gw1] linux2 Python 2.7.13 cwd: /home/dlezz/projects/robottelo-fork
[gw2] linux2 Python 2.7.13 cwd: /home/dlezz/projects/robottelo-fork
[gw3] linux2 Python 2.7.13 cwd: /home/dlezz/projects/robottelo-fork
[gw4] linux2 Python 2.7.13 cwd: /home/dlezz/projects/robottelo-fork
[gw5] linux2 Python 2.7.13 cwd: /home/dlezz/projects/robottelo-fork
[gw6] linux2 Python 2.7.13 cwd: /home/dlezz/projects/robottelo-fork
[gw7] linux2 Python 2.7.13 cwd: /home/dlezz/projects/robottelo-fork
[gw6] Python 2.7.13 (default, Feb  1 2017, 15:29:46)  -- [GCC 4.8.4]
[gw2] Python 2.7.13 (default, Feb  1 2017, 15:29:46)  -- [GCC 4.8.4]
[gw0] Python 2.7.13 (default, Feb  1 2017, 15:29:46)  -- [GCC 4.8.4]
[gw1] Python 2.7.13 (default, Feb  1 2017, 15:29:46)  -- [GCC 4.8.4]
[gw4] Python 2.7.13 (default, Feb  1 2017, 15:29:46)  -- [GCC 4.8.4]
[gw3] Python 2.7.13 (default, Feb  1 2017, 15:29:46)  -- [GCC 4.8.4]
[gw7] Python 2.7.13 (default, Feb  1 2017, 15:29:46)  -- [GCC 4.8.4]
[gw5] Python 2.7.13 (default, Feb  1 2017, 15:29:46)  -- [GCC 4.8.4] 
gw0 [6] / gw1 [6] / gw2 [6] / gw3 [6] / gw4 [6] / gw5 [6] / gw6 [6] / gw7 [6]
scheduling tests via LoadScheduling

tests/foreman/cli/test_host.py::HostCreateTestCase::test_positive_list_smartvariables_by_id <- robottelo/decorators/__init__.py 
tests/foreman/cli/test_host.py::HostCreateTestCase::test_positive_create_with_name 
tests/foreman/cli/test_host.py::HostCreateTestCase::test_positive_create_using_libvirt_without_mac <- robottelo/decorators/__init__.py 
tests/foreman/cli/test_host.py::HostCreateTestCase::test_positive_list_scparams_by_id <- robottelo/decorators/__init__.py 
tests/foreman/cli/test_host.py::HostCreateTestCase::test_positive_list_smartvariables_by_name <- robottelo/decorators/__init__.py 
tests/foreman/cli/test_host.py::HostCreateTestCase::test_positive_list_scparams_by_name <- robottelo/decorators/__init__.py 
[gw7] SKIPPED tests/foreman/cli/test_host.py::HostCreateTestCase::test_positive_create_using_libvirt_without_mac <- robottelo/decorators/__init__.py 
[gw0] PASSED tests/foreman/cli/test_host.py::HostCreateTestCase::test_positive_list_smartvariables_by_id <- robottelo/decorators/__init__.py 
[gw2] PASSED tests/foreman/cli/test_host.py::HostCreateTestCase::test_positive_list_scparams_by_name <- robottelo/decorators/__init__.py 
[gw6] PASSED tests/foreman/cli/test_host.py::HostCreateTestCase::test_positive_list_smartvariables_by_name <- robottelo/decorators/__init__.py 
[gw4] PASSED tests/foreman/cli/test_host.py::HostCreateTestCase::test_positive_list_scparams_by_id <- robottelo/decorators/__init__.py 
[gw3] PASSED tests/foreman/cli/test_host.py::HostCreateTestCase::test_positive_create_with_name 

======================================== 5 passed, 1 skipped in 172.87 seconds =========================================
```
**without the shared function the same tests have results:  5 passed, 1 skipped in 500.40 seconds** 
**if we count only the time gain using shared : 65%**

===============================================================================

test  api.test_host.HostTestCase 

```python
class HostTestCase(APITestCase):
    """Tests for ``entities.Host().path()``."""

    @classmethod
    @shared(inject=True)
    def setUpClass(cls, org_id=None, loc_id=None, cv_id=None, env_id=None,
                   lce_id=None, puppet_classes=None, image_id=None,
                   compresource_libvirt_id=None, _injected=False):
        """Setup common entities."""
        super(HostTestCase, cls).setUpClass()
        if _injected:
            cls.org = entities.Organization(id=org_id).read()
            cls.loc = entities.Location(id=loc_id).read()
            cls.cv = entities.ContentView(id=cv_id).read()
            cls.env = entities.Environment(id=env_id).read()
            cls.lce = entities.LifecycleEnvironment(id=lce_id).read()
            cls.puppet_classes = [entities.PuppetClass(id=cp_id) .read()
                                  for cp_id in puppet_classes]
            cls.compresource_libvirt = entities.LibvirtComputeResource(
                id=compresource_libvirt_id).read()
            cls.image = entities.Image(
                id=image_id, compute_resource=cls.compresource_libvirt).read()
        else:
            cls.org = entities.Organization().create()
            cls.loc = entities.Location(organization=[cls.org]).create()
            # Content View and repository related entities
            cls.cv = publish_puppet_module(
                [{'author': 'robottelo', 'name': 'generic_1'}],
                CUSTOM_PUPPET_REPO,
                organization_id=cls.org.id
            )
            cls.env = entities.Environment().search(
                query={'search': u'content_view="{0}"'.format(cls.cv.name)}
            )[0].read()
            cls.lce = entities.LifecycleEnvironment(
                organization=cls.org.id).search(query={
                    'search': 'name={0} and organization_id={1}'.format(
                        ENVIRONMENT, cls.org.id)
                }
            )[0].read()
            cls.puppet_classes = entities.PuppetClass().search(query={
                'search': u'name ~ "{0}" and environment = "{1}"'.format(
                    'generic_1', cls.env.name)
            })
            # Compute Resource related entities
            cls.compresource_libvirt = entities.LibvirtComputeResource(
                organization=[cls.org],
                location=[cls.loc],
            ).create()
            cls.image = entities.Image(
                compute_resource=cls.compresource_libvirt).create()
        puppet_classes = [pc.id for pc in cls.puppet_classes]
        return dict(org_id=cls.org.id, loc_id=cls.loc.id, cv_id=cls.cv.id,
                    env_id=cls.env.id, lce_id=cls.lce.id,
                    puppet_classes=puppet_classes,
                    compresource_libvirt_id=cls.compresource_libvirt.id,
                    image_id=cls.image.id)
```
test logs:

```console
(sat-6.2.z) dlezz@elysion:~/projects/robottelo-fork$ py.test tests/foreman/api/test_host.py -n 8 -v -k "HostTestCase"
================================================= test session starts ==================================================
platform linux2 -- Python 2.7.13, pytest-2.9.2, py-1.4.32, pluggy-0.3.1 -- /home/dlezz/.pyenv/versions/sat-6.2.z/bin/python
cachedir: .cache
rootdir: /home/dlezz/projects/robottelo-fork, inifile: 
plugins: xdist-1.15.0, services-1.1.14, cov-2.3.1
[gw0] linux2 Python 2.7.13 cwd: /home/dlezz/projects/robottelo-fork
[gw1] linux2 Python 2.7.13 cwd: /home/dlezz/projects/robottelo-fork
[gw2] linux2 Python 2.7.13 cwd: /home/dlezz/projects/robottelo-fork
[gw3] linux2 Python 2.7.13 cwd: /home/dlezz/projects/robottelo-fork
[gw4] linux2 Python 2.7.13 cwd: /home/dlezz/projects/robottelo-fork
[gw5] linux2 Python 2.7.13 cwd: /home/dlezz/projects/robottelo-fork
[gw6] linux2 Python 2.7.13 cwd: /home/dlezz/projects/robottelo-fork
[gw7] linux2 Python 2.7.13 cwd: /home/dlezz/projects/robottelo-fork
[gw4] Python 2.7.13 (default, Feb  1 2017, 15:29:46)  -- [GCC 4.8.4]
[gw6] Python 2.7.13 (default, Feb  1 2017, 15:29:46)  -- [GCC 4.8.4]
[gw1] Python 2.7.13 (default, Feb  1 2017, 15:29:46)  -- [GCC 4.8.4]
[gw5] Python 2.7.13 (default, Feb  1 2017, 15:29:46)  -- [GCC 4.8.4]  
[gw7] Python 2.7.13 (default, Feb  1 2017, 15:29:46)  -- [GCC 4.8.4]   
[gw3] Python 2.7.13 (default, Feb  1 2017, 15:29:46)  -- [GCC 4.8.4]    
[gw0] Python 2.7.13 (default, Feb  1 2017, 15:29:46)  -- [GCC 4.8.4]     
[gw2] Python 2.7.13 (default, Feb  1 2017, 15:29:46)  -- [GCC 4.8.4]      
gw0 [57] / gw1 [57] / gw2 [57] / gw3 [57] / gw4 [57] / gw5 [57] / gw6 [57] / gw7 [57]
scheduling tests via LoadScheduling

tests/foreman/api/test_host.py::HostTestCase::test_negative_update_os <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_comment <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_inherit_lce_cv <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_negative_update_arch <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_negative_update_mac <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_build_parameter <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_compresource <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_negative_update_name <- robottelo/decorators/__init__.py ^[^[
[gw6] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_build_parameter <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_image <- robottelo/decorators/__init__.py 
[gw5] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_inherit_lce_cv <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_hostgroup <- robottelo/decorators/__init__.py 
[gw4] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_compresource <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_managed_parameter <- robottelo/decorators/__init__.py 
[gw1] PASSED tests/foreman/api/test_host.py::HostTestCase::test_negative_update_os <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_host_parameters <- robottelo/decorators/__init__.py 
[gw7] PASSED tests/foreman/api/test_host.py::HostTestCase::test_negative_update_arch <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_compute_profile <- robottelo/decorators/__init__.py 
[gw6] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_image <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_model <- robottelo/decorators/__init__.py 
[gw1] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_host_parameters <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_provision_method <- robottelo/decorators/__init__.py 
[gw4] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_managed_parameter <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_owner_type <- robottelo/decorators/__init__.py 
[gw5] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_hostgroup <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_name <- robottelo/decorators/__init__.py 
[gw2] PASSED tests/foreman/api/test_host.py::HostTestCase::test_negative_update_mac <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_content_view <- robottelo/decorators/__init__.py 
[gw7] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_compute_profile <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_puppet_ca_proxy <- robottelo/decorators/__init__.py 
[gw6] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_model <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_puppet_class <- robottelo/decorators/__init__.py 
[gw0] PASSED tests/foreman/api/test_host.py::HostTestCase::test_negative_update_name <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_enabled_parameter <- robottelo/decorators/__init__.py 
[gw2] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_content_view <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_usergroup <- robottelo/decorators/__init__.py 
[gw1] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_provision_method <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_puppet_proxy <- robottelo/decorators/__init__.py 
[gw6] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_puppet_class <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_get_per_page <- robottelo/decorators/__init__.py 
[gw4] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_owner_type <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_subnet <- robottelo/decorators/__init__.py 
[gw7] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_puppet_ca_proxy <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_delete <- robottelo/decorators/__init__.py 
[gw0] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_enabled_parameter <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_get_search <- robottelo/decorators/__init__.py 
[gw0] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_get_search <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_compresource <- robottelo/decorators/__init__.py 
[gw1] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_puppet_proxy <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_read_puppet_proxy_name <- robottelo/decorators/__init__.py 
[gw6] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_get_per_page <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_arch <- robottelo/decorators/__init__.py 
[gw1] SKIPPED tests/foreman/api/test_host.py::HostTestCase::test_positive_read_puppet_proxy_name <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_content_view <- robottelo/decorators/__init__.py 
[gw2] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_usergroup <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_read_puppet_ca_proxy_name <- robottelo/decorators/__init__.py 
[gw4] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_subnet <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_build_parameter <- robottelo/decorators/__init__.py 
[gw2] SKIPPED tests/foreman/api/test_host.py::HostTestCase::test_positive_read_puppet_ca_proxy_name <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_env <- robottelo/decorators/__init__.py 
[gw7] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_delete <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_comment <- robottelo/decorators/__init__.py 
[gw1] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_content_view <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_enabled_parameter <- robottelo/decorators/__init__.py 
[gw0] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_compresource <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_compute_profile <- robottelo/decorators/__init__.py 
[gw6] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_arch <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_domain <- robottelo/decorators/__init__.py 
[gw5] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_name <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_user <- robottelo/decorators/__init__.py 
[gw2] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_env <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_hostgroup <- robottelo/decorators/__init__.py 
[gw5] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_user <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_medium <- robottelo/decorators/__init__.py 
[gw4] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_build_parameter <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_host_parameters <- robottelo/decorators/__init__.py 
[gw0] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_compute_profile <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_mac <- robottelo/decorators/__init__.py 
[gw6] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_domain <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_managed_parameter <- robottelo/decorators/__init__.py 
[gw3] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_comment <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_ip <- robottelo/decorators/__init__.py 
[gw2] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_hostgroup <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_model <- robottelo/decorators/__init__.py 
[gw4] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_host_parameters <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_os <- robottelo/decorators/__init__.py 
[gw1] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_enabled_parameter <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_ip <- robottelo/decorators/__init__.py 
[gw0] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_mac <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_owner_type <- robottelo/decorators/__init__.py 
[gw3] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_create_with_ip <- robottelo/decorators/__init__.py 
[gw5] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_medium <- robottelo/decorators/__init__.py 
[gw2] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_model <- robottelo/decorators/__init__.py 
[gw1] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_ip <- robottelo/decorators/__init__.py 
[gw4] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_os <- robottelo/decorators/__init__.py 
[gw0] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_owner_type <- robottelo/decorators/__init__.py 
[gw6] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_managed_parameter <- robottelo/decorators/__init__.py 
[gw7] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_comment <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_subnet <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_puppet_class <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_puppet_proxy <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_user <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_usergroup <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_puppet_ca_proxy <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_name <- robottelo/decorators/__init__.py 
tests/foreman/api/test_host.py::HostTestCase::test_positive_update_image <- robottelo/decorators/__init__.py 
[gw3] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_puppet_class <- robottelo/decorators/__init__.py 
[gw7] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_image <- robottelo/decorators/__init__.py 
[gw1] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_user <- robottelo/decorators/__init__.py 
[gw2] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_puppet_proxy <- robottelo/decorators/__init__.py 
[gw6] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_puppet_ca_proxy <- robottelo/decorators/__init__.py 
[gw4] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_subnet <- robottelo/decorators/__init__.py 
[gw5] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_name <- robottelo/decorators/__init__.py 
[gw0] PASSED tests/foreman/api/test_host.py::HostTestCase::test_positive_update_usergroup <- robottelo/decorators/__init__.py 

======================================== 55 passed, 2 skipped in 291.17 seconds ========================================
```
without the shared the test results was:  55 passed in 511.71 seconds 
**time gain 43%**

data file created location:  /tmp/robottelo/shared_functions/9c6af4d0664266ba2f0f7931c28eaf2d.shared_function.tests.foreman.api.test_host.HostTestCase.setUpClass 

the key: "9c6af4d0664266ba2f0f7931c28eaf2d" is md5 of the downloaded katello certificate of the server

data file content:
```json
{"state": "READY", "creation_datetime": "2017-03-24T13:22:13", "result": {"compresource_libvirt_id": 39, "loc_id": 767, "puppet_classes": [1], "cv_id": 449, "org_id": 766, "lce_id": 401, "image_id": 27, "env_id": 380}, "error": null, "pid": 15985, "id": "e01c45d7840e42e1833915729ae3bc78"}
``` 

The latest setUpClass can be rewritten like, to return the to_json_dict of each nailgun entity 
```python
    @classmethod
    @shared(inject=True)
    def setUpClass(cls, org=None, loc=None, cv=None, env=None,
                   lce=None, puppet_classes=None, image=None,
                   compresource_libvirt=None, _injected=False):
        """Setup common entities."""
        super(HostTestCase, cls).setUpClass()
        if _injected:
            # the returned kwargs are dicts restored from to_json_dict of
            # each entity except for puppet_classes that is ids list
            cls.org = entities.Organization().read(attrs=org)
            cls.loc = entities.Location().read(attrs=loc)
            cls.cv = entities.ContentView().read(attrs=cv)
            cls.env = entities.Environment().read(attrs=env)
            cls.lce = entities.LifecycleEnvironment().read(attrs=lce)
            cls.puppet_classes = [entities.PuppetClass(id=cp)
                                  for cp in puppet_classes]
            cls.compresource_libvirt = entities.LibvirtComputeResource().read(
                attrs=compresource_libvirt)
            # the image need an explicit read from server
            cls.image = entities.Image(
                id=image['id'],
                compute_resource=cls.compresource_libvirt).read()
        else:
            cls.org = entities.Organization().create()
            cls.loc = entities.Location(organization=[cls.org]).create()
            # Content View and repository related entities
            cls.cv = publish_puppet_module(
                [{'author': 'robottelo', 'name': 'generic_1'}],
                CUSTOM_PUPPET_REPO,
                organization_id=cls.org.id
            )
            cls.env = entities.Environment().search(
                query={'search': u'content_view="{0}"'.format(cls.cv.name)}
            )[0].read()
            cls.lce = entities.LifecycleEnvironment(
                organization=cls.org.id).search(query={
                    'search': 'name={0} and organization_id={1}'.format(
                        ENVIRONMENT, cls.org.id)
            }
            )[0].read()
            cls.puppet_classes = entities.PuppetClass().search(query={
                'search': u'name ~ "{0}" and environment = "{1}"'.format(
                    'generic_1', cls.env.name)
            })
            # Compute Resource related entities
            cls.compresource_libvirt = entities.LibvirtComputeResource(
                organization=[cls.org],
                location=[cls.loc],
            ).create()
            cls.image = entities.Image(
                compute_resource=cls.compresource_libvirt).create()

        return dict(org=cls.org, loc=cls.loc, cv=cls.cv, env=cls.env,
                    lce=cls.lce, compresource_libvirt=cls.compresource_libvirt,
                    puppet_classes=[pc.id for pc in cls.puppet_classes],
                    image=cls.image)
```
with this lasted modification the results was :  55 passed, 2 skipped in 279.71 seconds 
and time gain is 45%

shared file contain:
```json
{"state": "READY", "creation_datetime": "2017-03-27T10:35:49", "result": {"loc": {"subnet": [], "domain": [], "description": null, "smart_proxy": [], "compute_resource": [], "id": 1683, "environment": [], "config_template": [{"id": 7}, {"id": 8}, {"id": 9}, {"id": 41}, {"id": 10}, {"id": 11}, {"id": 14}, {"id": 13}, {"id": 12}, {"id": 72}, {"id": 71}, {"id": 42}, {"id": 43}, {"id": 15}, {"id": 16}, {"id": 82}, {"id": 44}, {"id": 45}, {"id": 17}, {"id": 18}, {"id": 19}, {"id": 46}, {"id": 20}, {"id": 47}, {"id": 76}, {"id": 21}, {"id": 22}, {"id": 23}, {"id": 39}, {"id": 37}, {"id": 38}, {"id": 26}, {"id": 28}, {"id": 27}, {"id": 29}, {"id": 48}, {"id": 40}, {"id": 30}, {"id": 31}, {"id": 33}, {"id": 32}, {"id": 34}, {"id": 49}, {"id": 50}, {"id": 4}, {"id": 5}, {"id": 6}, {"id": 2}, {"id": 3}, {"id": 1}, {"id": 51}, {"id": 52}, {"id": 53}, {"id": 77}, {"id": 73}, {"id": 75}, {"id": 74}, {"id": 69}, {"id": 35}, {"id": 36}], "user": [], "media": [], "organization": [{"id": 1682}], "hostgroup": [], "name": "nXKdpMIf"}, "puppet_classes": [1], "image": {"username": "\ud85d\udd2f\u7caa\u4433\u96d2\ud6bc\ud84d\udea6\ub28d\ucfa1\uc17a\u7839\ua03f\ud86d\udcf0\ud864\udd25\u9b6a\u0f59\ud846\udc81\u44ad\u424c\ud85e\udd70\ud861\udd5e\ud854\udd2f\u3d01\u37f6\ua22e\u9fb1\ud85f\udc52", "uuid": "\u85d0\ud802\udf24\ud866\udffd\u790d\u044d\u3d11\ud862\udf12\ud84b\udf1a\u9845\uc692\u7aa4", "compute_resource": {"provider_friendly_name": "Libvirt", "url": "http://xhedhJrraY.gov", "name": "f0QWO1RBuuz", "location": [{"id": 1683}], "provider": "Libvirt", "organization": [{"id": 1682}], "display_type": "spice", "id": 75, "set_console_password": true, "description": null}, "operatingsystem": {"id": 978}, "architecture": {"id": 977}, "id": 41, "name": "CWZwqv"}, "compresource_libvirt": {"provider_friendly_name": "Libvirt", "url": "http://xhedhJrraY.gov", "name": "f0QWO1RBuuz", "location": [{"id": 1683}], "provider": "Libvirt", "organization": [{"id": 1682}], "display_type": "spice", "id": 75, "set_console_password": true, "description": null}, "lce": {"organization": {"id": 1682}, "prior": null, "description": null, "name": "Library", "id": 866}, "env": {"organization": [{"id": 1682}], "location": [{"id": 2}], "id": 917, "name": "KT_iJrMFepW_Library_CKvnAyTVcf_924"}, "org": {"subnet": [], "domain": [], "description": null, "default_content_view": {"id": 923}, "smart_proxy": [{"id": 1}], "user": [], "title": "iJrMFepW", "hostgroup": [], "label": "iJrMFepW", "environment": [], "compute_resource": [], "config_template": [{"id": 7}, {"id": 8}, {"id": 9}, {"id": 41}, {"id": 10}, {"id": 11}, {"id": 14}, {"id": 13}, {"id": 12}, {"id": 72}, {"id": 71}, {"id": 42}, {"id": 43}, {"id": 15}, {"id": 16}, {"id": 82}, {"id": 44}, {"id": 45}, {"id": 17}, {"id": 18}, {"id": 19}, {"id": 46}, {"id": 20}, {"id": 47}, {"id": 76}, {"id": 21}, {"id": 22}, {"id": 23}, {"id": 39}, {"id": 37}, {"id": 38}, {"id": 26}, {"id": 28}, {"id": 27}, {"id": 29}, {"id": 48}, {"id": 40}, {"id": 30}, {"id": 31}, {"id": 33}, {"id": 32}, {"id": 34}, {"id": 49}, {"id": 50}, {"id": 4}, {"id": 5}, {"id": 6}, {"id": 2}, {"id": 3}, {"id": 1}, {"id": 51}, {"id": 52}, {"id": 53}, {"id": 77}, {"id": 73}, {"id": 75}, {"id": 74}, {"id": 69}, {"id": 35}, {"id": 36}], "library": {"id": 866}, "media": [], "id": 1682, "name": "iJrMFepW"}, "cv": {"puppet_module": [{"id": 63}], "description": null, "repository": [], "composite": false, "component": [], "last_published": "2017-03-27 10:35:30 UTC", "label": "CKvnAyTVcf", "version": [{"id": 921}], "organization": {"id": 1682}, "id": 924, "next_version": 2, "name": "CKvnAyTVcf"}}, "error": null, "pid": 28341, "id": "2422861e89fd4de8b4d703d43d733266"}
```
--
Also wanted to make test with subscription upload in setUpClass but using the use Katello agent that are failing any way.

ldjebran commented:
@renzo, by default we use storage=file even when configuration is missing, when infra will be ready with redis we can set storage=redis in config.

rochacbruno commented on Makefile:
```diff
@@ -139,6 +139,10 @@ logs-clean:
 	$(info "Removing pytest worker logs...")
 	-rm -f robottelo_gw*.log
 
+clean-shared:
```
It is missing on `PHONY` list at https://github.com/SatelliteQE/robottelo/blob/6.2.z/Makefile#L172

rochacbruno commented on requirements.txt:
```diff
@@ -5,12 +5,14 @@ click==6.6
 cryptography==1.4
 fauxfactory==2.0.9
 Inflector==2.0.11
+import_string==0.1.0
```
do we need to install it? AFAIK it is dependency of other module `manage` and manage installs its itself.

ldjebran replied:
@rochacbruno, we are using this library here in robotello by import,  think it should be as a dependency. if it's installed by other packages that's ok, this way other packages are free to change their dependencies.

ldjebran replied:
@rochacbruno , what do you think?

rochacbruno commented on robottelo/decorators/func_shared/file_storage.py:
```diff
@@ -0,0 +1,100 @@
+        return file_lock(self.get_key_file_path(lock_key), remove=False,
+                         timeout=self._lock_timeout)
+
+    def when_lock_acquired(self, handler):
```
Only a note:  https://docs.python.org/2.7/library/shelve.html can be used to have a filebase db.

ldjebran replied:
1 - I like the pickle functionality, but we need a more readable and basic to handle format for debugging in case of trouble 
2 - shelve here is useless as any way we need to implement a locking mechanism over keys
3 - if I had to choose a python pickle mechanism I much prefer zodb, that have transactions and a locking mechaniz can be easily implemented https://github.com/zopefoundation/ZODB , but for multi processes usage we have to start the ZEO service.

rochacbruno commented on robottelo/decorators/func_shared/shared.py:
```diff
@@ -0,0 +1,514 @@
+
+from robottelo.config import settings
+from robottelo.decorators.func_shared.file_storage import FileStorageHandler
+from robottelo.decorators.func_shared.redist_storage import RedisStorageHandler
```
why is it called `redist_storage` and not `redis_storage` intentional or typo?

ldjebran replied:
this is a typo :) thanks

abalakh commented on requirements.txt:
```diff
@@ -5,12 +5,14 @@ click==6.6
 pylint==1.6.4
 pytest==2.9.2
 pytest_services==1.1.14
+redis==2.10.5
```
Why adding redis as a requirement? For local runs we should be able to use file lock instead, so we should not insist on installing it. Imo it's a good candidate for requirements-optional.txt, but not requirements.txt

ldjebran replied:
ok will move it to, requirements-optional.txt

ldjebran commented:
@abalakh @rochacbruno @renzon made some requested updates now redis is very silent :)
please comment if updates needed

ldjebran commented:
@abalakh , @renzon @rochacbruno  after discussion with Andrey some changes has been made
- a configuration use_shared_data , by default false, this will force the shared function to always run the function
- at test start up log the shared_function config parameter use_shared_data 
- added clean-shared to clean-all 

now test boot looks like notice shared_function config: use result data (pre-cached): ON:
```console
platform linux2 -- Python 2.7.13, pytest-3.0.7, py-1.4.32, pluggy-0.4.0 -- /home/dlezz/.pyenv/versions/sat-6.2.z/bin/python
cachedir: .cache
shared_function config: use result data (pre-cached): ON
rootdir: /home/dlezz/projects/robottelo-fork, inifile:
plugins: xdist-1.15.0, services-1.1.14, cov-2.3.1
collected 22 items 
2017-04-13 17:11:11 - conftest - DEBUG - Found WONTFIX in decorated tests ['1156555', '1269196', '1402826', '1245334', '1221971', '1217635', '1226425', '1199150', '1204686', '1267224', '1103157', '1230902', '1214312', '1079482']

2017-04-13 17:11:11 - conftest - DEBUG - Collected 1 test cases
```

ldjebran commented:
@abalakh please do not merge until @rochacbruno and @renzon takes a look, thanks

renzon commented on robottelo/decorators/func_shared/redis_storage.py:
```diff
@@ -0,0 +1,72 @@
+    REDIS_DEFAULT_HOST = settings.shared_function.redis_host
+    REDIS_DEFAULT_HOST_PORT = settings.shared_function.redis_port
+    REDIS_DEFAULT_DB = settings.shared_function.redis_db
+    REDIS_DEFAULT_PASSWORD = settings.shared_function.redis_password
```
Be careful. Unfortunately calling configure on setting on import time cause some errors. @elyezer have pointed it out on some PRs

abalakh commented:
We've decided to discuss some details on Tue on our auto meeting, pls don't merge it until then.

oshtaier commented:
It is not like I just get to that PR right now, but discussing with Andrii, listening on the meetings and trying to understand benefits of that approach
Basically NACK as it is introduce unnecessary complexity into the code, decrease maintainability, increase chance of failure and etc. I don't like the idea to support these changes in the future, check whether some caches were cleared, that we are adding one possible root cause for any failure that we can get in future, that whole run can fail because I forgot to put necessary flags in some states and etc. Of course, if we get some dramatic improvement in performance, everything can be discussed once more, but in current implementation it should not get into master.
@ldjebran Could you please provide exact metrics of that improvement (like before these changes module run for 2 hours, and now 10 minutes)?

rochacbruno commented:
@oshtaier the point is that we need a solution to #4442 and future issues of the same type that may occur in future. 

For me, it looks like the proposed approach is trying to cover all possible scenarios (which is good) but we need to keep in mind that we have to fix the #4442 first with the simplest possible solution and then we may extend the implementation to cover other scenarios.

oshtaier commented:
@rochacbruno Sure. Doesn't seem that we have simplest solution there. It is just issue with multi-threading or multiprocessing, but we started to dig into caches, local files, additional servers and etc.

ldjebran commented:
@oshtaier @rochacbruno @renzon @abalakh 
lastest modified test is  tests/foreman/cli/test_host.py::KatelloAgentTestCase that current is running in one thread
```python
 @classmethod
    @skip_if_not_set('clients', 'fake_manifest')
    @shared(inject=True)
    def setUpClass(cls, org=None, env=None, content_view=None,
                   activation_key=None, _injected=False):
        """Create Org, Lifecycle Environment, Content View, Activation key

        """
        super(KatelloAgentTestCase, cls).setUpClass()
        if _injected:
            cls.org = org
            cls.env = env
            cls.content_view = content_view
            cls.activation_key = activation_key
            return

        # Create new org, environment, CV and activation key
        cls.org = make_org()
        cls.env = make_lifecycle_environment({
            u'organization-id': cls.org['id'],
        })
        cls.content_view = make_content_view({
            u'organization-id': cls.org['id'],
        })
        cls.activation_key = make_activation_key({
            u'lifecycle-environment-id': cls.env['id'],
            u'organization-id': cls.org['id'],
        })
        if settings.cdn:
            # Add subscription to Satellite Tools repo to activation key
            setup_org_for_a_rh_repo({
                u'product': PRDS['rhel'],
                u'repository-set': REPOSET['rhst7'],
                u'repository': REPOS['rhst7']['name'],
                u'organization-id': cls.org['id'],
                u'content-view-id': cls.content_view['id'],
                u'lifecycle-environment-id': cls.env['id'],
                u'activationkey-id': cls.activation_key['id'],
            })
        else:
            # Create custom internal Tools repo, add to activation key
            setup_org_for_a_custom_repo({
                u'url': settings.sattools_repo,
                u'organization-id': cls.org['id'],
                u'content-view-id': cls.content_view['id'],
                u'lifecycle-environment-id': cls.env['id'],
                u'activationkey-id': cls.activation_key['id'],
            })
        # Create custom repo, add subscription to activation key
        setup_org_for_a_custom_repo({
            u'url': FAKE_0_YUM_REPO,
            u'organization-id': cls.org['id'],
            u'content-view-id': cls.content_view['id'],
            u'lifecycle-environment-id': cls.env['id'],
            u'activationkey-id': cls.activation_key['id'],
        })
        return dict(org=cls.org, env=cls.env, content_view=cls.content_view,
                    activation_key=cls.activation_key)
```

in one thread : 9 passed, 1 skipped in 1559.46 seconds
with shared:  9 passed, 1 skipped in 615.71 seconds

ldjebran commented:
@oshtaier @rochacbruno @renzon @abalakh
time recap of all tests

| test                                                                                      |  without shared| with shared |
| ---------------------------------------------------------------------------- |---------------------|-----------------|
| tests/foreman/cli/test_host.py::HostCreateTestCase          |             500.40 |      172.87    |
| tests/foreman/cli/test_host.py::KatelloAgentTestCase        |           1559.46 |      615.71    |
| tests/foreman/api/test_host.py::HostTestCase                    |             511.71 |      291.17    |

ldjebran commented:
last changes

* removed the check settings from module levels and set it centralized from a function in shared module
* if shared enabled=false the function is run immidiately and transparently without any storage or looking
* scope problem resolution:
now shared function use scope in settings if not found will look at the environment variable ROBOTTELO_SHARED_FUNCTION_SCOPE and if not found by prevention will generate a new uuid 

    1. (This is the best scenario as it allow a better isolation) The robottelo-ci can generate a uuid and set it as the scope in properties file
    2. robottelo-ci can generate a uuid and export it via: 
        export ROBOTTELO_SHARED_FUNCTION_SCOPE=$(uuidgen)
       when master starts the log file this will like:
       2017-04-24 11:39:55 - conftest - DEBUG - got shared function scope from environment - dc202732ea314d39940de2a532a08fdb
    3. in case robottelo-ci has not generated/set scope for properties file and have not exported ROBOTTELO_SHARED_FUNCTION_SCOPE
       the master instance of pytest will generate a new uuid hex and create an environment variable ROBOTTELO_SHARED_FUNCTION_SCOPE
       when master starts the log file this will like:
       2017-04-24 11:39:55 - conftest - DEBUG - set environment shared function scope - dc202732ea314d39940de2a532a08fdb

* logging: 
    1. in case shared function enabled:
	at the start of the session log the shared usage in case enabled:
	shared_function enabled - ON - scope: 9bb50a012eab4f7b8f8c9f36934cf570 - storage: file
	or in case of redis:
	shared_function enabled - ON - scope: dc202732ea314d39940de2a532a08fdb - storage: redis

    2. in case shared function disabled:
       shared_function enabled - OFF - scope: None - storage: redis

last test run of katello agent tests:
 with storage=file http://pastebin.test.redhat.com/477696
 with storage=redis http://pastebin.test.redhat.com/477699

ldjebran commented:
last changes

* removed the check settings from module levels and set it centralized from a function in shared module
* if shared enabled=false the function is run immidiately and transparently without any storage or looking
* scope problem resolution:
experimented with generating a new scope and sharing it via enviroment but unfortunately this sollution will make more confusion between versions 6.3 and 6.2 running at the same time.

    1. The robottelo-ci can generate a uuid ($(uuidgen)) and set it as the scope in properties file 
    2. if scope was not defined in properties file a combinasion or server certificate md5 and parent process id will be used as a scope
       (this will give more propability that the scope is unique) 


* logging: 
    1. in case shared function enabled:
	at the start of the session log the shared usage in case enabled:
	shared_function enabled - ON - scope:  - storage: file
	or in case of redis:
	shared_function enabled - ON - scope:  - storage: redis

    2. in case shared function disabled:
       shared_function enabled - OFF - scope:  - storage: redis

last test run of katello agent tests:
 with storage=file http://pastebin.test.redhat.com/477787
 with storage=redis http://pastebin.test.redhat.com/477789

ldjebran commented:
@abalakh think have addressed all your comments, please confirm or comment.

ldjebran commented:
put back [DO NOT MERGE] label to be sure that all team members agree the changes before this PR get merged

rochacbruno commented:
As @lpramuk opened the issue #4442 it is important to have his review on this, specially to see if this PR fixes the needed feature of having 'upload manifest once'. Can you add you review, comments @lpramuk ?

ldjebran commented:
@abalakh I updated to use the parent master process id but to minimize the risk that this process is used again with other tests in other context I added the the md5 of the server certificate,
in earlier versions we had only the md5 certificate now I added the parent process
for example for katelloagent tests above the key now looks like: 
```console
ce54346efcef2250e5ceb6361c628e7c-8156.shared_function.tests.foreman.cli.test_host.KatelloAgentTestCase.setUpClass
```
where ce54346efcef2250e5ceb6361c628e7c  is the server md5 
and 8156 is the master parent process for the workers
the function now looks like 
```python
def _get_default_scope():
    """Return the shared function default scope"""

    _check_config()

    if NAMESPACE_SCOPE:
        return NAMESPACE_SCOPE

    return '{0}-{1}'.format(_get_server_cert_md5(), os.getppid())
```

ldjebran commented:
@abalakh the md5 is calculated only once as it's cached at module level

ldjebran commented:
@abalakh I want to be sure that this key data is used for only this server build

abalakh commented:
@ldjebran aren't we fighting windmills here? What's the chance of the same pid for the next process? 1/32767? And the chance that pid will be the same and exactly the same test class is being executed? Chance of same pid, same test class and valid timestamp (not timed out yet)? Also, which way md5 of the server would help us in case tests are executed against the same server (use case of debugging tests)?

If we were going to completely remove the chance of random failures - we would even have to stop using `gen_string` helper, for example, as with 0.0000001% chance (usually once per year) it generates same name for 2 entities, making a test fail as entity with the same name already exists.  But we haven't introduced any kind of scopes for every single test, don't we?

Last but not the least - downloading certificate and calculating md5 will definitely increase the time needed to start test execution. I remember times when simple tests could be executed in less then a second, now, with extra dependencies, bz status resolution and other things to simply output 'hello world' i have to wait around 10 seconds to get things ready for the test. And this time won't decrease with extra properties check/env variables check/downloading certificate/calculating md5 for sure.

ldjebran commented:
@abalakh here is the stat of md5
```console
from robottelo.config import setting
Traceback (most recent call last):
  File "/home/dlezz/.pyenv/versions/sat-6.2.z/lib/python2.7/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-ec9468487abd>", line 1, in <module>
    from robottelo.config import setting
ImportError: cannot import name setting
from robottelo.config import settings
import time
from robottelo.decorators.func_shared.shared import _get_server_cert_md5 
def test_md5():
    t = time.time()
    _get_server_cert_md5()
    print '>> server md5 time:', time.time() - t
    
settings.configure()
test_md5()
>> server md5 time: 0.526799201965
test_md5()
>> server md5 time: 1.09672546387e-05
test_md5()
>> server md5 time: 0.000118970870972
test_md5()
>> server md5 time: 6.79492950439e-05

```
first time usage 0.5 second
ulterior usage  less than a milli second

ldjebran commented:
@abalakh time is not critical here, and yes (it can seems strange) I have encountered PID re-usage in the past, I want to avoid to use it as an ID alone

ldjebran commented:
@abalakh the other tiers  processes are not launched just after this one , hundred of  processes can be launched in between in Jenkins

ldjebran commented:
@abalakh  question:Also, which way md5 of the server would help us in case tests are executed against the same server (use case of debugging tests)? 

We have the chance that the data is still valid and exist in the server, but when used with other server this is already a 100% failure

abalakh commented:
@ldjebran even if we wanna consider the chance of the same pid, why not just use hostname as a prefix and that's it? My point is you're digging too deep and introducing too complex and slow checks.
If you think that's the case of next test run against the same server, with same process id, same test class and same timeout period (and data for some reason can't be shared) is likely to happen, then what's the stopping point? This way we can assume md5 collision may happen too, so let's also check ssh key, some timestamps, network map etc. We can continue overcomplicating things till the point of total absurd :)

ldjebran commented:
@abalakh , using the server host is also a good option (we already use that for locking), 
is this an option you can consider?

ldjebran commented:
@abalakh I mean hostname + ppid

abalakh commented:
> We have the chance that the data is still valid and exist in the server, but when used with other server this is already a 100% failure

My point was different. Assuming i'm running some destructive test suite with shared resources against the same server multiple times (i wanna debug some tests, for example) - md5 of server certificate won't isolate me at all. pid is what will help here. md5 of server certificate will just isolate me from other server, but server's hostname would do the trick too.

> @abalakh , using the server host is also a good option (we already use that for locking),
is this an option you can consider?

If that also means removal of all the scoping setting/extra env variable/downloading certificate/calculating md5 checksum logic then i'm totally fine with it.

Btw, why default timeout is set to 24 hours, we've changed the approach and now are trying to share resources only inside the single test run, which means only different threads of the same run inside the same class should share them. Usually it's a matter of minutes, for some complex tests it's maybe an hour, but i don't think one test class run in multiple threads can last for more, than let's say 4 hours. What do you think? Imo with too big timeout value we're just slightly increasing the chance of re-using resources by wrong thread.

ldjebran commented:
we  arrived to the point that we need to validate the data max timeout to 24 hours not more 
so if share_timeout is more than 86400 seconds will have a validation error, that way we can use the ppid as a scope alone  
[shared] share time out cannot be more than 86400 seconds (24 hours)
will make the changes and run KatelloAgent tests with an other approach without injected

ldjebran commented:
update to use PPID if no scope defined with max timeout (life time) 24 hours
modified the katello testcase with other the approach without using inject
```python
    @classmethod
    @shared
    def _shared_function(cls):
        # Create new org, environment, CV and activation key
        org = make_org()
        env = make_lifecycle_environment({
            u'organization-id': org['id'],
        })
        content_view = make_content_view({
            u'organization-id': org['id'],
        })
        activation_key = make_activation_key({
            u'lifecycle-environment-id': env['id'],
            u'organization-id': org['id'],
        })
        if settings.cdn:
            # Add subscription to Satellite Tools repo to activation key
            setup_org_for_a_rh_repo({
                u'product': PRDS['rhel'],
                u'repository-set': REPOSET['rhst7'],
                u'repository': REPOS['rhst7']['name'],
                u'organization-id': org['id'],
                u'content-view-id': content_view['id'],
                u'lifecycle-environment-id': env['id'],
                u'activationkey-id': activation_key['id'],
            })
        else:
            # Create custom internal Tools repo, add to activation key
            setup_org_for_a_custom_repo({
                u'url': settings.sattools_repo,
                u'organization-id': org['id'],
                u'content-view-id': content_view['id'],
                u'lifecycle-environment-id': env['id'],
                u'activationkey-id': activation_key['id'],
            })
        # Create custom repo, add subscription to activation key
        setup_org_for_a_custom_repo({
            u'url': FAKE_0_YUM_REPO,
            u'organization-id': org['id'],
            u'content-view-id': content_view['id'],
            u'lifecycle-environment-id': env['id'],
            u'activationkey-id': activation_key['id'],
        })
        return dict(org=org, env=env, content_view=content_view,
                    activation_key=activation_key)

    @classmethod
    @skip_if_not_set('clients', 'fake_manifest')
    def setUpClass(cls):
        """Create Org, Lifecycle Environment, Content View, Activation key

        """
        super(KatelloAgentTestCase, cls).setUpClass()

        shared_data = cls._shared_function()
        cls.org = shared_data['org']
        cls.env = shared_data['env']
        cls.content_view = shared_data['content_view']
        cls.activation_key = shared_data['activation_key']
```

http://pastebin.test.redhat.com/478326

abalakh commented:
Ok, no new comments/suggestions in a week, merging it.
